vm/ByteCodeTranslator/src/com/codename1/tools/translator/ByteCodeTranslator.java

```diff
@@ -476,18 +476,54 @@ private static String getFileType(String s) {
         }
         return "file";
     }
-    
-    private static void replaceInFile(File sourceFile, String... values) throws IOException {
+    //
+    // Be kind to the GC; read as a StringBuilder, a data type designed
+    // to be mutated. Also, expire the temporary byte[] buffer so it can
+    // be collected.
+    //
+    private static StringBuilder readFileAsStringBuilder(File sourceFile) throws IOException
+    {
         DataInputStream dis = new DataInputStream(new FileInputStream(sourceFile));
         byte[] data = new byte[(int)sourceFile.length()];
         dis.readFully(data);
         dis.close();
-        FileWriter fios = new FileWriter(sourceFile);
-        String str = new String(data);
-        for(int iter = 0 ; iter < values.length ; iter += 2) {
-            str = str.replace(values[iter], values[iter + 1]);
+        StringBuilder b = new StringBuilder(new String(data));
+        return b;
+    }
+    //
+    // rewrite 4/2017 by ddyer to use more appropriate data
+    // structures, minimizing gc thrashing.  This avoids a big
+    // spike in memory and gc usage (and corresponding build 
+    // failures due to OutOfMemoryError) at the very end of the build 
+    // process for large projects.  
+    //
+    private static void replaceInFile(File sourceFile, String... values) throws IOException {
+    	StringBuilder str =readFileAsStringBuilder(sourceFile);
+    	int totchanges = 0;
+    	// perform the mutations on stringbuilder, which ought to implement
+    	// these operations efficiently.
+        for(int iter = 0 ; iter < values.length ; iter += 2) 
+        {
+        	String target = values[iter];
+        	String replacement = values[iter + 1];
+           	int found = 0;
+           	int index = 0;
+           	while((index=str.indexOf(target))>=0)
+           	{	int targetSize = target.length();
+           		str.replace(index, index+targetSize,replacement);
+           		index += replacement.length()-targetSize;
+           		found++;
+           		totchanges++;
+           	}
+           	if(found==0)
+        		{throw new Error("target string "+target+" is missing"); }
```

Comment:
It can't infinite loop - this is a single replacement and the next index starts after the 
replacement size.

Throwing the error is an element of style.  If the source doesn't contain the expected target,
would you rather be informed here, or by some undefined error down the line becuase the
file you constructed lacks something you intended to put there.